riscv/mmu.h

```diff
@@ -157,12 +167,14 @@ class mmu_t
         proc->state.log_mem_write.push_back(std::make_tuple(addr, (T)f(lhs), sizeof(T)));
       }
 
-      if (unlikely(!tlb_load_hit))
-        check_triggers(triggers::OPERATION_LOAD, addr, reg_from_bytes(sizeof(T), (uint8_t*)&res));
-      if (unlikely(!tlb_store_hit)) {
-        target_endian<T> target_val = to_target((T)f(lhs));
-        check_triggers(triggers::OPERATION_STORE, addr, reg_from_bytes(sizeof(T), (const uint8_t*)&target_val));
-      }
+      convert_trigger_before_to_trigger_after({
```

Comment:
I don't understand why all data triggers on AMOs should be converted to `timing=1 (after)`. This does not seem compatible with the debug spec to me. The spec warns about this case for loads:

> Data load triggers with timing of 0 will result in the same load happening again when the debugger lets the hart run. For data load triggers, debuggers must first attempt to set the breakpoint with timing of 1.

It also says that `timing` is expected to be fixed by the hardware:

> Most hardware will only implement one timing or the other, possibly dependent on select, execute, load, and store. This bit primarily exists for the hardware to communicate to the debugger what will happen. Hardware may implement the bit fully writable, in which case the debugger has a little more control.

So it seems to me that if someone sets a data breakpoint with `timing=0 (before)`, they are taking a risk, and if it results in repeating the AMO, that's their problem.

If we wanted to avoid this footgun, we should have mcontrol triggers force `timing=1 (after)` when `select=1 (data)`.

Reply:
Agree that legalizing tdata1.timing works better.
I will try to implement a legalize_timing() function based on Table 5.10.
![image](https://user-images.githubusercontent.com/39526191/210477946-ff7a5015-b4a5-469b-a306-51acb4e59e7d.png)
(Clarifying https://github.com/riscv/riscv-debug-spec/issues/792)

Reply:
PR https://github.com/riscv-software-src/riscv-isa-sim/pull/1214 adds a legalize_timing() for legalizing tdata1.timing of load data trigger.